pkg/stanza/fileconsumer/file_sort.go

```diff
@@ -0,0 +1,243 @@
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"time"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.uber.org/multierr"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/timeutils"
+)
+
+const (
+	SortTypeNumeric      = "numeric"
+	SortTypeTimestamp    = "timestamp"
+	SortTypeAlphabetical = "alphabetical"
+)
+
+type SortRule interface {
+	validate() error
+	sort(re *regexp.Regexp, files []string) ([]string, error)
+}
+
+type BaseSortRule struct {
+	RegexKey  string `mapstructure:"regex_key,omitempty"`
+	Ascending bool   `mapstructure:"ascending,omitempty"`
+	SortType  string `mapstructure:"sort_type,omitempty"`
+}
+
+type SortRuleImpl struct {
+	SortRule
+}
+
+func (sr *SortRuleImpl) Unmarshal(component *confmap.Conf) error {
+	if !component.IsSet("sort_type") {
+		return fmt.Errorf("missing required field 'sort_type'")
+	}
+
+	typeInterface := component.Get("sort_type")
+
+	typeString, ok := typeInterface.(string)
+	if !ok {
+		return fmt.Errorf("non-string type %T for field 'sort_type'", typeInterface)
+	}
+
+	switch typeString {
+	case SortTypeNumeric:
+		var numericSortRule NumericSortRule
+		err := component.Unmarshal(&numericSortRule, confmap.WithErrorUnused())
+		if err != nil {
+			return err
+		}
+		sr.SortRule = numericSortRule
+	case SortTypeAlphabetical:
+		var alphabeticalSortRule AlphabeticalSortRule
+		err := component.Unmarshal(&alphabeticalSortRule, confmap.WithErrorUnused())
+		if err != nil {
+			return err
+		}
+		sr.SortRule = alphabeticalSortRule
+	case SortTypeTimestamp:
+		var timestampSortRule TimestampSortRule
+		err := component.Unmarshal(&timestampSortRule, confmap.WithErrorUnused())
+		if err != nil {
+			return err
+		}
+		sr.SortRule = timestampSortRule
+	default:
+		return fmt.Errorf("invalid sort type %s", typeString)
+	}
+
+	return nil
+}
+
+type NumericSortRule struct {
+	BaseSortRule `mapstructure:",squash"`
+}
+
+func (f NumericSortRule) validate() error {
+	if f.RegexKey == "" {
+		return fmt.Errorf("regex key must be specified for numeric sort")
+	}
+	return nil
+}
+
+type AlphabeticalSortRule struct {
+	BaseSortRule `mapstructure:",squash"`
+}
+
+func (f AlphabeticalSortRule) validate() error {
+	if f.RegexKey == "" {
+		return fmt.Errorf("regex key must be specified for alphabetical sort")
+	}
+	return nil
+}
+
+type TimestampSortRule struct {
+	BaseSortRule `mapstructure:",squash"`
+	Format       string `mapstructure:"format,omitempty"`
+	Location     string `mapstructure:"location,omitempty"`
+}
+
+func (f TimestampSortRule) validate() error {
+	if f.RegexKey == "" {
+		return fmt.Errorf("regex key must be specified for timestamp sort")
+	}
+	if f.Format == "" {
+		return fmt.Errorf("format must be specified for timestamp sort")
+	}
+	if f.Location == "" {
+		f.Location = "UTC"
+	}
+	_, err := time.Parse(f.Format, "")
+	if err != nil {
+		return fmt.Errorf("error parsing format %s: %w", f.Format, err)
+	}
+
+	_, err = time.LoadLocation(f.Location)
```

Comment:
Ideally we'd validate using the `timeutils` package, so that we're not applying different requirements.

Can you suggest necessary changes there?

Reply:
Updated to use the Getlocation and just to ignore format

Reply:
Was this actually done?

Reply:
shows as outdated for me? not seeing `loadlocation` anywhere in this package.